### PR TITLE
Hardware tutorial

### PR DESCRIPTION
**Pull request by josh146**
**Context:**

In #319, the changes were requested for the hardware tutorial.

**Description of the Change:**

- [x] In hardware tutorial add a download link for test.xbb.

  *Note: in the refactor, the file `test.xbb` is no longer used.*

- [x] Specify the INFILE argument to sf run is a blackbird script

- [x] Add thumbnail for the hardware tutorial

  image

- [x] In hardware tutorial, link readers back to the operation decorator docstring

- [x] Can we restructure this section so that it foremost shows users how to rescale their adjacency matrix (if possible) to give a runnable circuit?

  *Note: this is non-trivial, so have removed this discussion from the tutorial. I recommend we add in a function that automatically rescales the adjacency matrix for hardware chips.*

- [x] Also, is the value r=1 an actual physical constraint? I thought it was just binary 

- [x] thewalrus appears quite suddenly here. Suggest writing a smoother segue.
  
  *Note: have removed mention of thewalrus*.

- [x] Enough to just reproduce the error message, not the full traceback (which anyway contains some unnecessary developer-specific stuff)

**Benefits:** n/a

**Possible Drawbacks:** n/a

**Related GitHub Issues:** #319

### CODE REVIEW

File: doc/introduction/remote.rst [hunk 1]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
-layout of the hardware device, you can apply any four-mode interferometer to the pairs of modes.
+        with prog.context as q:
+            ops.S2gate(1.0) | (q[0], q[4])
+            ops.S2gate(1.0) | (q[1], q[5])
+            ops.S2gate(1.0) | (q[2], q[6])
+            ops.S2gate(1.0) | (q[3], q[7])
 
-Primitive gates supported by ``chip2`` include any combination of:
+            unitary() | q[:4]
+            unitary() | q[4:]
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
            unitary() | q[4:]
 
            ops.MeasureFock() | q
```

File: doc/introduction/remote.rst [hunk 2]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+-----------------------
 
-* `rotations/phase shifts <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.Rgate.html>`_ (:class:`~.ops.Rgate`).
+In the above example, ``eng.run()`` is a **blocking** method; Python will wait until
+the job result has been returned from the cloud before further lines
+will execute.
 
-Furthermore, several automatic decompositions are supported:
+**Non-blocking jobs** can also be submitted using the :meth:`eng.run_async() <~.RemoteEngine.run_async>` method. Unlike ``eng.run()``, this method returns a :class:`~.Job`
+object that can be queried to get the jobs status.
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
object that can be queried to get the job's status.
```

File: doc/introduction/remote.rst [hunk 3]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+>>> job.result
+[[0 1 0 2 1 0 0 0]]
 
-    # Define a unitary matrix
-    complex array U[4, 4] =
-         0.09980516-0.78971535j,  0.53374613+0.07984545j, -0.21161788+0.10047649j, -0.01337026-0.14167555j
-         -0.12759979-0.00425289j,  0.14089156+0.40091225j, 0.31942372-0.21453252j, -0.79775306+0.13657774j
-         -0.18224807+0.30281836j,  0.26930442-0.04644871j, -0.46045639-0.55359506j, -0.0737605-0.52580999j
-         0.19903677-0.43076659j, -0.50320649-0.44750373j, -0.01617065-0.52755812j, -0.19729219+0.06200712j
+Finally, an incomplete job can be *cancelled* by calling :meth:`job.cancel() <~.Job.cancel>`.
```
[the remark this answers is not in the document]
Comment (trbromley) [reply doc/introduction/remote.rst]: Although, the method link didn't work for me again.

File: doc/introduction/remote.rst [hunk 4]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
-    BSgate(0.543, -0.123) | [0, 1]
+Hardware compilation
+--------------------
 
-    # Duplicate the above unitary for
-    # the second pair of modes
-    Interferometer(U) | [4, 5, 6, 7]
-    BSgate(0.543, -0.123) | [4, 5]
+When creating a quantum program to run on hardware, Strawberry Fields allows any
+combination of the following gates when constructing the unitary to be applied:
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
When creating a quantum program to run on hardware, Strawberry Fields can compile
any collection of the following gates into a multi-mode unitary:
```
Just felt this sentence read a bit strangely.

File: doc/introduction/remote.rst [hunk 5]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+* You can use the :class:`~.ops.Interferometer` command to directly pass a
+  unitary matrix to be decomposed and compiled to match the device architecture.
+  This performs a rectangular decomposition using Mach-Zehnder interferometers.
 
-This program will execute following the same steps as above; :class:`~.RemoteEngine` will automatically
-compile the program to match the layout of the chip.
+* You can use :class:`~.ops.BipartiteGraphEmbed` to embed a bipartite graph on
+  the GBS chip. Note, however, that the decomposed squeezing values depends on the graph
+  structure, so only bipartite graphs that result in equal squeezing on all
+  modes can currently be executed on currently available chips.
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
  modes can be executed on currently available chips.
```

File: doc/introduction/remote.rst [hunk 6]
```diff
@@ -352,52 +288,101 @@ Rgate(1.902) | (q[0])
...
+    # the first pair of modes, as well
+    # as a beamsplitter
+    Interferometer(A0) | [0, 1, 2, 3]
+    BSgate(0.543, 0.123) | [2, 0]
+    Rgate(0.453) | 1
+    MZgate(0.65, -0.54) | [2, 3]
+
+    # Duplicate the above unitary for
+    # the second pair of modes
+    Interferometer(A1) | [4, 5, 6, 7]
```
Comment (trbromley) [inline doc/introduction/remote.rst]: Interesting that it prefers to use another identical matrix `A1` instead of going for `A0` again.

File: doc/introduction/remote.rst [hunk 7]
```diff
@@ -352,52 +288,101 @@ Rgate(1.902) | (q[0])
...
+.. math:: A = \begin{bmatrix}0 & B\\ B^T & 0\end{bmatrix}
 
-The matrix :math:`B` represents the edges between the two sets of
-vertices in the graph, and :math:`A` is the full adjacency matrix
-:math:`A = \begin{bmatrix}0 & B\\ B^T & 0\end{bmatrix}`. Here, we will
-consider a complete bipartite graph, since we know that the singular
-values are of the form :math:`\{d, 0\}`.
+where :math:`B` represents the edges between the two sets of
+vertices in the graph. However, the devices are currently restricted
+to bipartite graphs such that the singular values form the set :math:`\{0, d\}`
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
to bipartite graphs with equally sized partitions, such that the singular values form the set :math:`\{0, d\}`
```
Not sure though :thinking:

File: doc/introduction/remote.rst [hunk 8]
```diff
@@ -352,52 +288,101 @@ Rgate(1.902) | (q[0])
...
+for some real value :math:`d`.
+
+Here, we will
+consider a `complete bipartite graph <https://en.wikipedia.org/wiki/Complete_bipartite_graph>`_,
+since the singular values are of the form :math:`\{d, 0\}`.
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
since the singular values are of the form :math:`\{0, d\}`.
```
Just to match the one above.

File: doc/introduction/remote.rst [hunk 9]
```diff
@@ -443,68 +428,15 @@ Rgate(-0.8273) | (q[0])
...
-        141
-
-
-    CircuitError: Incorrect squeezing value(s) (r, phi)={(1.336, 0.0), (0.177, 0.0), (0.818, 0.0)}. Allowed squeezing value(s) are (r, phi)={(1, 0.0), (0.0, 0.0)}.
+>>> prog = sf.Program(8)
+>>> with prog.context as q:
+...     ops.BipartiteGraphEmbed(A, mean_photon_per_mode=1) | q
+...     ops.MeasureFock() | q
+CircuitError: Incorrect squeezing value(s) (r, phi)={(1.336, 0.0), (0.177, 0.0), (0.818, 0.0)}.
+Allowed squeezing value(s) are (r, phi)={(1, 0.0), (0.0, 0.0)}.
```
Comment (trbromley) [inline doc/introduction/remote.rst]: I'm not sure if this is that useful for the tutorial. If a user did run into the error message, it is quite self explanatory.

File: strawberryfields/circuitspecs/gaussian_unitary.py [hunk 10]
```diff
@@ -22,40 +22,41 @@
...
     """Compiler to arrange a Gaussian quantum circuit into the canonical Symplectic form.
+
     This compile specification checks whether the circuit can be implemented as a sequence of
     Gaussian operations. If so, it arranges them in the canonical order with displacement at the end.
     After compilation, the circuit will consist of at most two operations, a :class:`~.GaussianTransform`
     and a :class:`~.Dgate`.
-    This compiler can be accessed by calling :meth:`.Program.compile` with `'gaussian_unitary'` specified.
 
-        **Example:**
+    This compiler can be accessed by calling :meth:`.Program.compile` with `'gaussian_unitary'` specified.
```
Comment (trbromley) [inline strawberryfields/circuitspecs/gaussian_unitary.py]: Was this a case of lowering the indentation?

File: doc/introduction/remote.rst [hunk 11]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
 
+We can also use the Python collections module to convert the samples into
+counts:
 
-We can convert the samples into counts using the following function:
+>>> from collections import Counter
+>>> bitstrings = [tuple(i) for i in results.samples]
+>>> counts = {k:v for k, v in Counter(bitstrings).items()}
+>>> counts[(0, 0, 0, 0, 0, 0, 0)]
+2
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/remote.rst]: Yep, people were asking how to do it.

File: doc/introduction/remote.rst [hunk 12]
```diff
@@ -197,17 +123,17 @@ Next we create the program
 
         ops.MeasureFock() | q
 
-We create the engine. The engine is in charge of compiling and executing
-programs on the remote device.
+Finally, we create the engine. Similarly to the :class:`~.LocalEngine`, the :class:`~.RemoteEngine`
+is in charge of compiling and executing programs, however it differs in that the program will be
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
is in charge of compiling and executing programs. However, it differs in that the program will be
```

File: doc/introduction/remote.rst [hunk 13]
```diff
@@ -197,17 +123,17 @@ Next we create the program
...
+executed on *remote* devices, rather than on local simulators.
 
->>> eng = RemoteEngine("chip2")
+>>> eng = RemoteEngine("X8_01")
 
 We run the engine by calling ``eng.run``, and pass it the program we
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: Better to phrase as running the program rather than running the engine

File: doc/introduction/remote.rst [hunk 14]
```diff
@@ -228,104 +154,122 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
        [0, 0, 0, 0, 0, 1, 0, 0],
        [0, 0, 1, 1, 0, 2, 1, 2],
        [2, 0, 1, 0, 1, 0, 0, 0]])
->>> np.mean(results.samples.T, axis=0)
+
+The samples returned correspond to 20 measurements (or shots) of the 8 mode quantum program
+above. While the majority of modes have measured zero photons, we can see that a couple have
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
above. Some modes have measured zero photons, and others have
```

File: doc/introduction/remote.rst [hunk 15]
```diff
@@ -228,104 +154,122 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+detected single photons, with a few even detecting 2 or 3.
+
+By taking the average of the returned array along the shots axis, we can estimate the
+mean photon number of each mode:
+
+>>> np.mean(results.samples, axis=0)
 array([0.4 , 0.1 , 0.15, 0.05, 0.3 , 0.3 , 0.45, 0.35])
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
array([0.4, 0.1, 0.15, 0.05, 0.3, 0.3, 0.45, 0.35])
```

File: doc/introduction/remote.rst [hunk 16]
```diff
@@ -228,104 +154,122 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
 
-In addition to using the program template above, which directly matches the physical
-layout of the hardware device, you can apply any four-mode interferometer to the pairs of modes.
+        with prog.context as q:
+            ops.S2gate(1.0) | (q[0], q[4])
+            ops.S2gate(1.0) | (q[1], q[5])
+            ops.S2gate(1.0) | (q[3], q[7])
 
-Primitive gates supported by ``chip2`` include any combination of:
+            unitary() | q[:4]
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: Still find the switching of signatures here weird

File: doc/introduction/remote.rst [hunk 17]
```diff
@@ -228,104 +154,122 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+Managing submitted jobs
+-----------------------
 
-* `rotations/phase shifts <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.Rgate.html>`_ (:class:`~.ops.Rgate`).
+In the above example, :meth:`RemoteEngine.run` is a **blocking** method; Python will wait until
+the job result has been returned from the cloud before further lines
+will execute.
 
-Furthermore, several automatic decompositions are supported:
+**Non-blocking jobs** can also be submitted using the :meth:`eng.run_async() <~.RemoteEngine.run_async>` method. Unlike ``eng.run()``, this method returns a :class:`~.Job`
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: Suggest rewording. It's not really the job that's blocking here, it's the call

File: doc/introduction/remote.rst [hunk 18]
```diff
@@ -228,104 +154,122 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+Furthermore, several automatic decompositions are supported:
 
-You may wish to view the compiled program; this can be easily done in Python using
-the ``Program.print`` method:
+* You can use the :class:`~.ops.Interferometer` command to directly pass a
+  unitary matrix to be decomposed and compiled to match the device architecture.
+  This performs a rectangular decomposition using Mach-Zehnder interferometers.
 
+* You can use :class:`~.ops.BipartiteGraphEmbed` to embed a bipartite graph on
+  the GBS chip. Note, however, that the decomposed squeezing values depends on the graph
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: First time GBS is mentioned, I believe

Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
  the GBS chip. Note, however, that the decomposed squeezing values depend on the graph
```

Comment (co9olguy) [inline doc/introduction/remote.rst]: better to actually make a callout Note here

File: doc/introduction/remote.rst [hunk 19]
```diff
@@ -228,104 +154,122 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+  modes can be executed on currently available chips.
 
->>> from strawberryfields import RemoteEngine
->>> from strawberryfields.io import load
->>> prog = load("test.xbb")
->>> prog = prog.compile("chip2")
->>> prog.print()
+Before sending the program to the cloud platform to be executed, however, Strawberry Fields
+must **compile** the program to match the physical architecture or layout of the photonic chip.
+This happens implicitly when using the remote engine, however we can use the :meth:`~.Program.compile`
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: Does the engine recompile already-compiled programs?

File: doc/introduction/remote.rst [hunk 20]
```diff
@@ -352,52 +296,101 @@ Rgate(1.902) | (q[0])
...
->>> from strawberryfields.io import save
->>> save("test_compiled.xbb", prog)
 
+Working with Blackbird scripts
+------------------------------
 
-Tips and tricks
----------------
+When submitting quantum programs to be executed remotely, they are communicated to
+the cloud platform using Blackbird --- a quantum photonic assembly language.
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
the cloud platform using Blackbird---a quantum photonic assembly language.
```

File: doc/introduction/remote.rst [hunk 21]
```diff
@@ -352,52 +296,101 @@ Rgate(1.902) | (q[0])
...
+Strawberry Fields also supports exporting programs directly as Blackbird scripts
+(an ``xbb`` file); Blackbird scripts can then be submitted to be executed via the
+Strawberry Fields :doc:`command line interface </code/sf_cli>`.
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: Does this link render properly? (didn't check myself, but looks similar to ones @trbromley commented on)

File: doc/introduction/remote.rst [hunk 22]
```diff
@@ -352,52 +296,101 @@ Rgate(1.902) | (q[0])
...
+For example, lets save our program above as a Blackbird script in the
+current directory, via the :func:`~.save` function:
 
-    from strawberryfields.utils import operation
+>>> sf.save("program1.xbb", prog)
 
-We can define an operation to make it easier to apply the same unitary
-to both signal and idler modes.
+This produces the following Blackbird script, where comments have been added
+for clarity:
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
This produces the following Blackbird script (comments have been added
for clarity):
```

File: doc/introduction/remote.rst [hunk 23]
```diff
@@ -352,52 +296,101 @@ Rgate(1.902) | (q[0])
...
+
+.. code-block:: console
+
+    $ sf run program1.xbb --output out.txt
+
+After executing the above command, the result will be stored in ``out.txt`` in the
+current working directory. You can also omit the ``--output`` parameter to print the
+result to the screen.
+
+Furthermore, saved Blackbird scripts can be imported as a Strawberry Fields program
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
Furthermore, saved Blackbird scripts can be imported as Strawberry Fields programs
```

File: doc/introduction/remote.rst [hunk 24]
```diff
@@ -352,52 +296,101 @@ Rgate(1.902) | (q[0])
...
+.. math:: A = \begin{bmatrix}0 & B\\ B^T & 0\end{bmatrix}
+
+where :math:`B` represents the edges between the two sets of
+vertices in the graph. However, the devices are currently restricted
+to bipartite graphs with equally sized partitions, such that the singular values form the set :math:`\{0, d\}`
+for some real value :math:`d`.
+
+Here, we will
+consider a `complete bipartite graph <https://en.wikipedia.org/wiki/Complete_bipartite_graph>`_,
+since the singular values are of the form :math:`\{0, d\}`.
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: Didn't like my `zeros_like` suggestion? :crying_cat_face:

File: doc/introduction/remote.rst [hunk 25]
```diff
@@ -443,68 +436,15 @@ Rgate(-0.8273) | (q[0])
...
-        141
-
-
-    CircuitError: Incorrect squeezing value(s) (r, phi)={(1.336, 0.0), (0.177, 0.0), (0.818, 0.0)}. Allowed squeezing value(s) are (r, phi)={(1, 0.0), (0.0, 0.0)}.
+>>> prog = sf.Program(8)
+>>> with prog.context as q:
+...     ops.BipartiteGraphEmbed(A, mean_photon_per_mode=1) | q
+...     ops.MeasureFock() | q
+CircuitError: Incorrect squeezing value(s) (r, phi)={(1.336, 0.0), (0.177, 0.0), (0.818, 0.0)}.
+Allowed squeezing value(s) are (r, phi)={(1, 0.0), (0.0, 0.0)}.
```
Comment (co9olguy) [inline doc/introduction/remote.rst]: does the error message actually print (1, 0.0)? i.e., a combination of integer and float?

File: doc/introduction/remote.rst (hunk 6 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Yep, I didn't know this myself until now!

File: doc/introduction/remote.rst (hunk 24 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Oh, I wasn't sure what you meant by it! And then in the process of making the new PR, accidentally lost track of that comment. Did you mean write `A` using `zeros_like`?

File: strawberryfields/circuitspecs/gaussian_unitary.py (hunk 10 above)
Comment (josh146) [reply strawberryfields/circuitspecs/gaussian_unitary.py]: I think this was always indented incorrectly, and I only just noticed it!

File: doc/introduction/remote.rst [hunk 26]
```diff
@@ -228,104 +149,119 @@ array([[0, 0, 1, 0, 1, 0, 1, 0],
...
+Managing submitted jobs
+-----------------------
 
-* `rotations/phase shifts <https://strawberryfields.readthedocs.io/en/stable/code/api/strawberryfields.ops.Rgate.html>`_ (:class:`~.ops.Rgate`).
+In the above example, ``eng.run()`` is a **blocking** method; Python will wait until
+the job result has been returned from the cloud before further lines
+will execute.
 
-Furthermore, several automatic decompositions are supported:
+**Non-blocking jobs** can also be submitted using the :meth:`eng.run_async() <~.RemoteEngine.run_async>` method. Unlike ``eng.run()``, this method returns a :class:`~.Job`
```
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/remote.rst]: Ah, thanks for catching this Tom! I had a stray `~`, removing this makes it work.

File: doc/introduction/remote.rst (hunk 3 above)
[the remark this answers is not in the document]
Comment (josh146) [reply doc/introduction/remote.rst]: Fixed!

File: doc/introduction/remote.rst (hunk 13 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Rephrase this to:
> We can now run the program by calling ``eng.run``, and passing the program to be executed
as well as additional runtime options.

File: doc/introduction/remote.rst (hunk 16 above)
Comment (josh146) [reply doc/introduction/remote.rst]: This is a result of the `operation` decorator. I added this to the tutorial originally a couple of months ago, as several people were asking for the best shortcut to only define the unitary once - worth removing?

File: doc/introduction/remote.rst (hunk 17 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Rephrased to:
```rest
Jobs can also be submitted using the **non-blocking**
:meth:`eng.run_async() <.RemoteEngine.run_async>` method.
```

File: doc/introduction/remote.rst (hunk 18 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Ah, good catch. I've replaced it with 'photonic chip' for now. This tutorial is a bit lacking wrt the wider context at the moment, as we do not yet have (a) a introduction page to our photonic hardware, and (b) an introduction to GBS (existing page is aimed more at the advanced level)

File: doc/introduction/remote.rst (hunk 19 above)
Comment (josh146) [reply doc/introduction/remote.rst]: This is the `eng.run` logic:
```python
# if the program hasn't been compiled for this backend, do it now
target = self.backend.circuit_spec
if target is not None and p.target != target:
    p = p.compile(target, **compile_options)
```

So compilation only occurs if the existing program target does not match the engine backend.

File: doc/introduction/remote.rst (hunk 18 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Made a warning box 🙂 
image

File: doc/introduction/remote.rst (hunk 21 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Yep! 
image

File: doc/introduction/remote.rst (hunk 25 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Just double checked, it does. This is an artifact of the allowed squeezing value being an integer `1` in the chip spec, which is being programmatically inserted into the error message. We should fix this, but better to do it in a separate PR once the new HW chips have been added

File: doc/introduction/remote.rst (hunk 16 above)
Comment (co9olguy) [reply doc/introduction/remote.rst]: No, but worth rethinking the `operation` decorator. Possible to make it something like a `property` which can be called without brackets?

File: doc/introduction/remote.rst (hunk 24 above)
Comment (co9olguy) [reply doc/introduction/remote.rst]: Instead of using `0*B`, use `np.zeros_like(B)`

File: doc/introduction/remote.rst (hunk 25 above)
Comment (co9olguy) [reply doc/introduction/remote.rst]: k can we document so we don't forget?

File: doc/introduction/remote.rst [hunk 27]
```diff
@@ -200,20 +200,22 @@ counts:
...
+
             ops.MeasureFock() | q
 
     Refer to the :class:`~.operation` documentation for more details.
 
 Managing submitted jobs
 -----------------------
 
-In the above example, :meth:`RemoteEngine.run` is a **blocking** method; Python will wait until
+In the above example, :meth:`eng.run <.RemoteEngine.run>` is a **blocking** method; Python will wait until
```
Comment (co9olguy) [inline doc/introduction/remote.rst]:
```suggestion
In the above example, :meth:`eng.run() <.RemoteEngine.run>` is a **blocking** method; Python will wait until
```

File: doc/introduction/remote.rst (hunk 16 above)
Comment (josh146) [reply doc/introduction/remote.rst]: Hmm, I don't think so. It can actually accept (optional) arguments if required, so you can do something like this:

```python
@operation
custom_op(a, b, q):
    Dgate(a) | q[0]
    Rgate(b) | q[1]

with prog.context as q:
    custom_op(a, b) | q[0, 1]
```

However, the `q` argument always disappears from the signature, as the decorator moves it to the right of the `|` operator.

File: doc/introduction/remote.rst [hunk 28]
```diff
@@ -443,68 +442,15 @@ Rgate(-0.8273) | (q[0])
...
-
-Note that the above squeezing values must be of the form :math:`\{0,1\}`
-to be embedded on the chip. Consider a bipartite graph where this is not
-the case:
+If the bipartite graph to be embedded does not satisfy the aforementioned
+restriction on the singular values, an error message will be raised on
+compilation:
 
 >>> B = np.array([[0, 1, 0, 1], [1, 0, 1, 0], [0, 1, 1, 1], [1, 0, 1, 0]])
 >>> A = np.block([[0*B, B], [B.T, 0*B]])
```
Comment (josh146) [inline doc/introduction/remote.rst]:
```suggestion
>>> A = np.block([[np.zeros_like(B), B], [B.T, np.zeros_like(B)]])
```

File: doc/introduction/remote.rst (hunk 11 above)
[the remark this answers is not in the document]
Comment (trbromley) [reply doc/introduction/remote.rst]: Ok, if it helps people :smile:

File: doc/introduction/remote.rst [hunk 29]
```diff
@@ -183,29 +188,34 @@ counts:
...
 
-In the above example, ``eng.run()`` is a **blocking** method; Python will wait until
+In the above example, :meth:`eng.run() <.RemoteEngine.run>` is a **blocking** method; Python will wait until
 the job result has been returned from the cloud before further lines
 will execute.
 
-**Non-blocking jobs** can also be submitted using the :meth:`eng.run_async() <~.RemoteEngine.run_async>` method. Unlike ``eng.run()``, this method returns a :class:`~.Job`
-object that can be queried to get the jobs status.
+Jobs can also be submitted using the **non-blocking** :meth:`eng.run_async() <.RemoteEngine.run_async>`
+method. Unlike ``eng.run()``, which returns a :class:`~.Results` object once the computation is
```
Comment (trbromley) [inline doc/introduction/remote.rst]:
```suggestion
method. Unlike :meth:`eng.run() <.RemoteEngine.run>`, which returns a :class:`~.Results` object once the computation is
```

File: doc/introduction/remote.rst (hunk 16 above)
Comment (co9olguy) [reply doc/introduction/remote.rst]: Possible to put the `q` as an argument to the decorator rather than the function?

**Comment by co9olguy** [comment]
Can we add a `mailto` in a prominant place for people to use when they want to request hardware access. Address will be sf @ xanadu.ai

File: doc/introduction/remote.rst (hunk 16 above)
Comment (josh146) [reply doc/introduction/remote.rst]: I'll make an issue and we can look into it :+1:

Comment (josh146) [reply doc/introduction/remote.rst]: Ah, started writing out the issue, and released it's not possible that way. The function is a _definition_, `q` is an undefined dummy variable, whereas in the decorator you are calling an already defined function with a `q` that must be defined.

File: doc/introduction/remote.rst [hunk 30]
```diff
@@ -1,180 +1,106 @@
...
+* Compiling quantum programs for specific quantum chip architectures,
+* Saving and running Blackbird scripts, an assembly language for quantum photonic hardware, and
+* Embedding and sampling from graphs on photonic chips.
 
-1. By using the :func:`~.store_account` function to store your account credentials:
+.. warning::
 
-   >>> sf.store_account("AUTHENTICATION_TOKEN")
+    An authentication token is required to access the Xanadu cloud platform. In the
+    following, replace ``AUTHENTICATION_TOKEN`` with your personal API token.
```
Comment (josh146) [inline doc/introduction/remote.rst]:
```suggestion
    An authentication token is required to access the Xanadu cloud platform. In the
    following, replace ``AUTHENTICATION_TOKEN`` with your personal API token.
    
    If you do not have an authentication token, you can request hardware access via email
    at sf@xanadu.ai.
```
